pkg/disttask/framework/dispatcher/dispatcher.go

```diff
@@ -397,37 +401,114 @@ func (d *BaseDispatcher) replaceDeadNodesIfAny() error {
-				replaceNodes[nodeID] = disttaskutil.GenerateExecID(n.IP, n.Port)
-				cleanNodes = append(cleanNodes, nodeID)
-			}
+		d.LiveNodes = newInfos
+		// 3. rebalance subtasks.
+		d.LiveNodes, err = d.filterByRole(d.LiveNodes)
+		if err != nil {
+			return err
 		}
-		if len(replaceNodes) > 0 {
-			logutil.Logger(d.logCtx).Info("reschedule subtasks to other nodes", zap.Int("node-cnt", len(replaceNodes)))
-			if err := d.taskMgr.UpdateFailedSchedulerIDs(d.ctx, d.Task.ID, replaceNodes); err != nil {
-				return err
-			}
-			if err := d.taskMgr.CleanUpMeta(d.ctx, cleanNodes); err != nil {
-				return err
-			}
-			// replace local cache.
-			for k, v := range replaceNodes {
-				for m, n := range d.taskNodes {
-					if n == k {
-						d.taskNodes[m] = v
-						break
-					}
-				}
+
+		if len(d.LiveNodes) > 0 {
+			return d.RebalanceSubtasksImpl()
+		}
+		return nil
+	}
+	return nil
+}
+
+func (d *BaseDispatcher) replaceTaskNodes() {
+	d.TaskNodes = d.TaskNodes[:0]
+	for _, serverInfo := range d.LiveNodes {
+		d.TaskNodes = append(d.TaskNodes, disttaskutil.GenerateExecID(serverInfo.IP, serverInfo.Port))
+	}
+}
+
+// RebalanceSubtasksImpl rebalance subtasks from taskNodes to liveNodes.
+func (d *BaseDispatcher) RebalanceSubtasksImpl() error {
+	// 1. find out nodes that scaled out.
+	scaleOutNodes := make([]string, 0)
+	for _, node := range d.LiveNodes {
+		execID := disttaskutil.GenerateExecID(node.IP, node.Port)
+		if !disttaskutil.MatchSchedulerID(d.TaskNodes, disttaskutil.GenerateExecID(node.IP, node.Port)) {
+			scaleOutNodes = append(scaleOutNodes, execID)
+		}
+	}
+
+	// 2. find out nodes need to clean subtasks.
+	cleanNodes := make([]string, 0)
+	cleanNodeMap := make(map[string]bool, 0)
+	for _, node := range d.TaskNodes {
+		if !disttaskutil.MatchServerInfo(d.LiveNodes, node) {
+			cleanNodeMap[node] = true
+			cleanNodes = append(cleanNodes, node)
+		}
+	}
+
+	if len(scaleOutNodes) == 0 && len(cleanNodes) == 0 {
+		return nil
+	}
+
+	// 3. get subtasks for each node before scaling out.
+	subtasks, err := d.taskMgr.GetSubtasksByStepExceptStates(d.ctx, d.Task.ID, d.Task.Step, proto.TaskStateSucceed)
+	if err != nil {
+		return err
+	}
+
+	subtasksForEachScheduler := make(map[string][]*proto.Subtask, len(d.LiveNodes))
+	for _, subtask := range subtasks {
+		subtasksForEachScheduler[subtask.SchedulerID] = append(
+			subtasksForEachScheduler[subtask.SchedulerID],
+			subtask)
+	}
+
+	// 4. scale out subtasks to scaleOutNodes.
+	lastScaleOutIdx := 0
+	liveNodeSubtaskCnt := 0
+	for id, v := range subtasksForEachScheduler {
+		if ok := cleanNodeMap[id]; ok {
+			continue
+		}
+		liveNodeSubtaskCnt += len(v)
+	}
+
+	averageSubtaskCnt := liveNodeSubtaskCnt / len(d.LiveNodes)
+	for id, v := range subtasksForEachScheduler {
+		if ok := cleanNodeMap[id]; ok {
+			continue
+		}
+		if len(v) > averageSubtaskCnt {
+			for i := 0; i < len(v)-averageSubtaskCnt; i++ {
+				v[i].SchedulerID = scaleOutNodes[lastScaleOutIdx%len(scaleOutNodes)]
```

Comment:
this loop should consider together with below loop:

suppose we have 5 live nodes, 2 out of it is scale-out nodes, 1 dead node
on live node 1 we have 2 subtasks, on dead node 1 we have 3 subtasks
after this loop, we might have this"
1 subtask on node 1, 1 on some scale-out node

and after below loop, we have
2 1 1 on live nodes
1 0 on scale out nodes